### ISSUE TITLE
Allow initializing a stack that already exists

### ISSUE THREAD

**Issue by ffMathy** [pulumi/pulumi#11393]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

The title sounds a bit strange, so allow me to explain.

I have a setup where I (when starting my project locally) programmatically copy my development stack using `pulumi stack init local --non-interactive --copy-config-from dev --secrets-provider some-secrets-provider`.

From here, I then use Pulumi Automation to "up" the stack causing it to create a copy of my whole AWS environment, but in the *same* AWS account, but where every resource of this new "local" stack is prefixed with a particular string (could be the branch I am working on).

This works great, and allows me to have N amount of test environments - one per branch, in the same AWS account.

However, I'd also like to do this trick in my build runner. So whenever it builds my branch, I'd like it to also use `pulumi stack init local --non-interactive --copy-config-from dev --secrets-provider some-secrets-provider` and then run tests towards that AWS environment and deploy to it.

But this causes issues, because the stack already "exists" in S3 from when I created it locally (I use S3 as a state backend), but not locally on the disk, where the runner needs it to read configuration files from etc. Furthermore, if two separate runners then try to do this trick, they also lack the local stack files, while the stack exists in the backend.

Perhaps the ultimate solution to this would actually be that it would be possible to have one stack that has multiple *states*, and you could pick a state by name. That way, I could have multiple copies of my AWS resources with different states, but in the same stack.

### Affected area/feature

- Automation API
- Pulumi CLI

**Comment by ffMathy** [comment]
Might relate a bit (or add fuel to the fire) on the discussion over at #8402.

**Comment by aq17** [comment]
Thanks for the suggestion, @ffMathy – we'll add this to our backlog for consideration.
A workaround may be to deploy another stack with the same (or different) config as your development stack. As you stated that you'd like `multiple copies of my AWS resources with different states`, managing those different states with a separate stack should be best for now.

**Comment by ffMathy** [comment]
But isn't that what I am doing?

From what I understand, the init command copies the config from another stack?

**Comment by aq17** [comment]
The `--copy-config-from` flag does, but it sounds like you may be attempting to initialize an already existing stack `local` numerous times, rather than standing up a new distinct (but identical) stack?

**Comment by ffMathy** [comment]
Yes. Since I am copying from my "dev" stack which I already have locally checked in to GIT (so the stack config is there) and it also exists in S3 state-wise, I basically just want to be able to create a copy of the stack.

However, since the copy I am creating already exists in S3 (but not in GIT, the config file is not present on the build server, because it's a copy I made locally), I can't do it.

Would be awesome to just bypass the check and allow it to copy my config anyway.

So basically imagine the following scenario:

1. I am on my local machine. I programmatically copy the stack "dev" to the stack "local". This causes the stack config file to be created, and the S3 state populated with the new stack. The config file is *not* added to GIT, as this is a debug-stack that's bound to my branch name.
2. I then push my commit. The build server now tries to spin up the same stack. But since the state exists in S3 for the new stack, but the runner doesn't have the config file, it can't do it. It'd be awesome if I could just copy my "dev" stack to "local" again on the runner, and then don't care that the state already exists in S3. I just want the config copy.

The reason I can't just copy the file, is that it has a secrets provider (KMS).

Does that add more context that helps explain this?

**Comment by Frassle** [comment]
Little hesitant to allow init to not error for already existing stacks. Creating new stacks is the common case here and most users want an error if they accidentally try to create a stack that already exists.

It seems the main thing your wanting here is the config copied across, is `pulumi config cp` not sufficient for this:
```
$ pulumi config cp --help
Copies the config from the current stack to the destination stack. If `key` is omitted,
then all of the config from the current stack will be copied to the destination stack.
```

**Comment by ffMathy** [comment]
But copying a stack requires an existing stack to exist, no?

And the problem is that the stack can't be created, because it has already been created on another machine.

**Comment by ffMathy** [comment]
@Frassle is it clear what I mean by this, or should I try to re-elaborate? 🙂

**Comment by Frassle** [comment]
Ah yes sorry, I'd suggest trying to use `pulumi stack select --create` and `pulumi config cp`.

We could possibly add a `--select` flag to init (sort of mirror of the `--create` flag in select), which could re-init a stack if it already exists. 
There's quite a few questions around how exactly that would work though.
e.g. what would `pulumi stack init --select --secrets-provider=passphrase --stack dev` do if the stack dev already exists and is using a kms secret provider? What about if it was already using the passphrase secret provider, would it ask for a new password? What about any stack resources in the stack would it clear those?

**Comment by grantjoy** [comment]
The way I see it, it would be great if `pulumi stack init` operated as an upsert - I'm not seeing that functionality today.

**Comment by atlantiz8** [comment]
> Ah yes sorry, I'd suggest trying to use `pulumi stack select --create` and `pulumi config cp`.
> 
> We could possibly add a `--select` flag to init (sort of mirror of the `--create` flag in select), which could re-init a stack if it already exists. There's quite a few questions around how exactly that would work though. e.g. what would `pulumi stack init --select --secrets-provider=passphrase --stack dev` do if the stack dev already exists and is using a kms secret provider? What about if it was already using the passphrase secret provider, would it ask for a new password? What about any stack resources in the stack would it clear those?

pulumi stack select --create works well for automated pipeline, but problem is we'll have to manually select the old stack again in order to copy the config to newly created stack. Possible for us to have a `pulumi config copy-from` and it'll pull the config from the old stack instead ?
